dev/PushNotifications/PushNotificationManager.cpp

```diff
@@ -221,249 +247,365 @@ namespace winrt::Microsoft::Windows::PushNotifications::implementation
+        {
+            if (PushNotificationHelpers::IsPackagedAppScenario())
             {
-                GUID taskClsid = details.TaskClsid();
-                THROW_HR_IF(E_INVALIDARG, taskClsid == GUID_NULL);
-
                 {
-                    auto lock = s_activatorInfoLock.lock_shared();
-                    THROW_HR_IF(E_INVALIDARG, s_pushTriggerRegistration);
+                    auto lock{ m_lock.lock_shared() };
+                    THROW_HR_IF(HRESULT_FROM_WIN32(ERROR_ALREADY_REGISTERED), m_comActivatorRegistration || m_pushTriggerRegistration);
                 }
 
-                winrt::hstring taskClsidStr{ winrt::to_hstring(taskClsid) };
-                winrt::hstring backgroundTaskFullName{ backgroundTaskName + taskClsidStr };
-
-                auto tasks{ BackgroundTaskRegistration::AllTasks() };
-                bool isTaskRegistered{ std::any_of(std::begin(tasks), std::end(tasks),
-                    [&](auto&& task)
+                auto scopeExitToCleanRegistrations{ wil::scope_exit([&]()
+                {
                     {
-                        auto name{ task.Value().Name() };
+                        auto lock { m_lock.lock_exclusive() };
+                        m_comActivatorRegistration.reset();
 
-                        if (std::wstring_view(name).substr(0, backgroundTaskName.size()) != backgroundTaskName)
+                        // Clean the task registration only if it was created during this call
+                        if (m_pushTriggerRegistration)
                         {
-                            return false;
+                            m_pushTriggerRegistration.Unregister(true);
+                            m_pushTriggerRegistration = nullptr;
                         }
 
-                        if (name == backgroundTaskFullName)
-                        {
-                            s_pushTriggerRegistration = task.Value();
-                            return true;
-                        }
+                        m_registering = false;
+                    }
 
-                        throw winrt::hresult_invalid_argument(L"RegisterActivator has different clsid registered.");
-                    }) };
+                    auto appUserModelId{ PushNotificationHelpers::GetAppUserModelId() };
+                    LOG_IF_FAILED(PushNotifications_UnregisterNotificationSinkForFullTrustApplication(appUserModelId.get()));
+                }
+                )};
 
-                if (!isTaskRegistered)
+                winrt::hstring backgroundTaskFullName;
                 {
-                    builder = BackgroundTaskBuilder();
+                    auto lock{ m_lock.lock_shared() };
+                    backgroundTaskFullName = backgroundTaskName + winrt::to_hstring(m_registeredClsid);
+                }
+
+                // Register a PushTrigger for BI to activate the application through COM
+                if (!IsBackgroundTaskRegistered(backgroundTaskFullName))
```

Comment:
If the background task is already registered by a previous instance of the process, then what? m_pushTriggerRegistration would be nullptr due to this check

Reply:
m_pushTriggerRegistration was being set in IsBackgroundTaskRegistered. Making another helper function to retrieve the backgroundTaskRegistration for more transparency.